### PR TITLE
TextEditingDelta Support for the Web

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
## Description
[Design Doc](https://docs.google.com/document/d/1PXNd_LwJudiUTnCM2AXRxbd98_p4BlFf6qFhEhv3wvs/edit?usp=sharing)

This change aims to collect the changes made to the editing value in the web text input plugin, and send them to the framework through a platform channel.

Framework PR: flutter/flutter#88477

## Related Issues
Partially fixes flutter/flutter#87972

## Tests
Added tests to verify delta inference.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:464]: Would it make this any simpler if you used [inputType](https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/inputType)?  Or is that not possible...

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:478]: Nit: Line length.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:461]: Why are the names slightly different than the frameworks' TextDelta, is it to match how the browser's events present the information?

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:809]: I think it makes sense to add this parameter rather than creating a new type as long as nothing is broken by it.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:493]: Is this whole comment explaining the situation of double space inserting a period?  Maybe add that as an example in the comment somewhere.  And/or, be sure that this is thoroughly tested so that future developers can understand your intention (and not break it).

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:505]: So in the example of double space inserting a period, the matches here will be all periods in the text, and you're trying to find the one that is new?

**Review by justinmc — COMMENTED** [review]
I mainly want to make sure that there's no simpler way to do `inferDeltaState`.  I've left some questions on there.  Glad to hear it's working though.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:505]: in the case of double space inserting a period the `deltaText` we receive from `beforeinput` is actually `. `. So we search for a period followed by a whitespace in the new editing state. The new editing state comes directly from the `DOM` so it can be seen as the source of truth. 

When we have the matched ranges we apply iterate through them and apply the `deltaText` to the `oldText` using these matched ranges as the `deltaRange`. The range that results in our `oldText` being equal to the value of the new editing state is the correct `deltaRange` that we should send to the framework.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:461]: I just tried to match the `EditingState` parameters in this case.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:493]: Add a test for this case and the accent menu case.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1193]: Maybe what you need is `js_util.getProperty(event, 'data') as String?`? That way you can avoid comparing with the string `'null'` below.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:998]: If I understand correctly, you are using `lastTextEditingDeltaState` to accumulate all the delta information during the sequence of events (`beforeinput`, `compositionupdate`, `input`, etc) before flushing the change to the framework. If that's the case, I have a few suggestions:

1. Make the  `TextEditingDeltaState` class mutable (i.e. remove `final` from its fields) so you can update the fields in-place instead of creating many copies.
2. Remove the `last` prefix, it's confusing because it doesn't mean the same thing as `last` in `lastEditingState`.
3. Since we don't know what's the best time to create this mutable object, I suggest doing something like this:
```dart
TextEditingDeltaState? _editingDelta;
TextEditingDeltaState get editingDelta {
  if (_editingDelta == null) {
    _editingDelta = TextEditingDeltaState(oldText: lastEditingState!.text!);
  }
  return _editingDelta!;
}
```
Then you can simply do:
```dart
editingDelta.deltaStart = ...
editingDelta.deltaText = ...
```
And when the change has been flushed to the framework, just set the delta to null:
```dart
_editingDelta = null;
```

Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:998]: Hey @mdebbar thank you for the reviews! I wasn't able to get to this in the past few days but I will come back to it and address all these comments. Thanks again!

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart]: I attempted to use `inputType` however there is some normalization I do in order to get the behavior of the Web to fit into the processing we do on the framework side that doesn't allow for this. 

For example, when trying to use `insertCompositionText` I discovered it is called for the entire composed word.
So say we have `hello|` and we delete character by character starting at the end. In this case `insertCompositionText` is called for every deletion. However on Android & iOS, when we delete the final character `h` this is reported as a regular deletion, and not one from the composing region.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:456]: It seems like the point of this class is to represent the json that will be sent to the framework, so maybe mention that in the comment.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:994]: I was confused at first by the fact that some TextEditingDeltaStates are named `delta` and some are `deltaState`.  What do you think of normalizing them all to `deltaState`?  So here `editingDelta` would become `editingDeltaState`.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:477]: Related to a comment I'm leaving below where inferDeltaState is called: What do you think of not modifying the lastTextEditingDelta argument?  Maybe just make a copy if that makes sense.

I'm not sure if this is in the Flutter style guide, but I personally try not to modify function arguments unless I have to.  Seeing this makes me think that the variable passed in as lastTextEditingDelta where inferDeltaState is called is expecting to be modified and will be reused for something.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1179]: Related to a comment I left above about modifying function arguments: Is editingDelta equivalent to newTextEditingDeltaState after this line?  If so, it would simplify things to not modify the function parameter and just modify a copy, and then to use newTextEditingDeltaState exclusively after this.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1185]: Did you mean to pass editingDelta here and not newTextEditingDeltaState? (Or per my comment above, maybe they are equivalent.)  I'm confused by the naming here in general.  Is it possible to make it more consistent with the naming for the non-delta variables here?  (Where lastEditingState becomes newEditingState and is passed to onChange).

**Review by justinmc — COMMENTED** [review]
I left some comments about adding doc comments and maybe improving naming in order to make this a bit easier to understand.  It's probably worth it to explain some of the flow that you explained to me on Discord in the comments.

I was trying to think of a better way to structure the code so that it's easier to follow the flow of putting together the final TextEditingDeltaState, but I couldn't come up with anything solid...  Maybe something will come to me after you fix some of the naming things I mentioned in my comments.  Otherwise it should be fine if it's just explained a bit more in the comments.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:994]: That makes sense to me. I'll do that .

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:477]: I see what you mean. I think it makes sense to not modify the lastTextEditingDelta argument since we are constructing a new delta with the new information we have at this point in the flow.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1179]: editingDelta is not equivalent to newTextEditingDeltaState after inferDeltaState completes. iirc I don't think dart supports pass by reference. I agree making a copy instead would simplify things here.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart]: At this point newTextEditingDeltaState is equivalent to editingDelta. On line 1184 we set _editingDelta to newTextEditingDeltaState, and then editingDelta that is passed to onChange is equivalent since it is a getter for _editingDelta.

Here I was trying to follow the same pattern the editingState uses. First getting the new editing state from the DOM; I get the new editingDelta from inferDeltaState. And then lastEditingState is set to the newEditingState; I set the _editingDelta to the newTextEditingDeltaState. And then both of these are passed to onChange. 

I normalized some of the names in a recent change. Let me know if it's still confusing.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1251]: This makes a lot more sense to me at a glance now, thanks for putting up with my pickiness.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:489]: Thanks for doing this, I think it's comforting to know there's nothing fancy going on with modifying the inputs when I see this copy.

**Review by justinmc — COMMENTED** [review]
I think the code looks great, but I tried running this and was able to crash it in two different ways:

  1. In the middle of the text, insert a bunch of spaces.  Sometimes the spaces don't seem to work, sometimes they insert a period (which I think should not happen on desktop web), and eventually it crashes.

<details>

<summary>Video of crash #1</summary>

https://user-images.githubusercontent.com/389558/151609835-2c725a2e-db5a-472b-8ab6-087d1ab693b9.mov

</details>

  2. Insert "ni" on the Chinese keyboard and replace it with 你.  It crashes.
 
<details>

<summary>Video of crash #2</summary>


https://user-images.githubusercontent.com/389558/151610092-58b00cf6-5be2-4714-8910-c517042057cb.mov

</details>

I'm using [my fork](https://github.com/justinmc/delta-text-editor-demo) of the rich text editing demo to test this.  There may also be problems with the demo that are contributing to this...

**Comment by Renzo-Olivares** [comment]
> I think the code looks great, but I tried running this and was able to crash it in two different ways:
> 
> 1. In the middle of the text, insert a bunch of spaces.  Sometimes the spaces don't seem to work, sometimes they insert a period (which I think should not happen on desktop web), and eventually it crashes.
> 
> Video of crash #1
> 2. Insert "ni" on the Chinese keyboard and replace it with 你.  It crashes.
> 
> Video of crash #2
> I'm using [my fork](https://github.com/justinmc/delta-text-editor-demo) of the rich text editing demo to test this. There may also be problems with the demo that are contributing to this...

1. I can confirm this issue, but inserting a period when doing a double space is native behavior on macOS/iOS (in between a word, and at the end of a word). On Android it will only insert a period at the end of a word, and not if you double space in between a word. Linux does not insert a period in either case.

2. Confirmed this is an issue with the demo.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart]: I don't think so. But I agree, I fixed it.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Thanks for looking at those two crashes.  The double space thing seems better to me now, but I hope to get time to really polish up that demo so that we can better test this kind of thing in a realistic environment.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:470]: Are the empty strings and -1 meaningful values, or are they here only to make the fields non-null? Would it make sense to mark the parameters with `required`?

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:478]: nit: s/out/our/ (although I'd prefer "the")
nit: backtick the `deltaStart`

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:495]: nit: consider extracting complex conditions into named variables. Less needs to be explained this way. E.g.:

```dart
final bool isTextBeingRemoved = newTextEditingDeltaState.deltaText.isEmpty && newTextEditingDeltaState.deltaEnd != -1;
if (isTextBeingRemoved) {
  ...
```

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:598]: If empty strings and -1 are meaningful values for these fields, let's explain what they mean. Something tells me that the 3 fields above could be combined into a `TextEditingDelta(text, start, end)` struct and the default here expressed as `TextEditingDelta? delta;`. I may be missing something though.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:542]: Yes it does contain unescaped text. I think a better solution here would be to use `RegExp.escape(deltaText)` which will give us an escaped version of the given text.

**Review by yjbanov — APPROVED** [review]
LGTM if LGT @justinmc

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:470]: `required` wouldn't work in this case since we don't create the entire `TextEditingDelta` at any single point. It goes through `beforeInput` -> `onInput` -> `inferDeltaState`.

Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart]: On the framework side `-1` for `deltaStart` and `deltaEnd` is currently used by `TextEditingDelta` to identify a `NonTextUpdate`.  I made the `selection` and `composing` nullable now.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:596]: I think since the selection and composing region change when the text is mutated they should be together with `deltaStart`,`deltaEnd`, and `deltaText` in one class.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:451]: Nit: Maybe just return this directly instead of assigning to a variable.

**Review by justinmc — APPROVED** [review]
Updated LGTM 👍 

I feel confident about merging this without a recent LGTM from Mouad, but we should check in with him when he's back.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart]: sgtm

**Pull request merged**
